# An Introduction to Combining Data with Pandas' `merge`, `join`, and `concat`

This notebook contains code examples to use with the article of the same name, along with light commentary on each of the examples. 

In [1]:
import pandas as pd

pd.set_option("display.max_columns", 50)

## Importing Data

Here you will import the temperature and precipitation climate normals datasets into DataFrames. Calling `.head()` on the DataFrame will give you a 5 row preview of your data, and the `shape` attribute will give you the dimensions of the data in the form `(rows, columns)`. These are great sanity checks to run before doing too much with the data. 

In [2]:
climate_temp = pd.read_csv("climate_temp.csv")
climate_temp.head()

STATION            STATION_NAME ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                6                7               10               15  
1                6                7               10               15  
2                5                7               10               15  
3                5                7               10               15  
4                5                7               10               15

In [3]:
climate_temp.shape

(127020, 21)

In [4]:
climate_precip = pd.read_csv("climate_precip.csv")
climate_precip.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6            -6.66             -666   
1             -666            -66.6            -6.66             -666   
2             -666            -66.6            -6.66             -666   
3            -9999          -9999.0            -6.66            -9999   
4            -9999          -9999.0            -6.66            -9999   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6            -6.66             -666            -66.6   
1            -66.6            -6.66             -666            -66.6   
2            -66.6            -6.66             -666            -66.6   
3          -9999.0            -6.66            -9999          -9999.0   
4          -9999.0            -6.66            -9999          -9999.0   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.02              0.0             0.02              0.0   
1             0.04              0.0             0.04              0.0   
2             0.05              0.0             0.05              0.0   
3             0.07              0.0             0.07              0.0   
4             0.09              0.0             0.09              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                       98                       43                       12   
1                       99                       44                       12   
2                      100                       44                       12   
3                      101                       45                       12   
4                      102                       45                       12   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                        3                    -9999                        0   
1                        3                    -9999                        0   
2                        3                    -9999                        0   
3                        3                        0                        0   
4                        3                        0                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                    -9999                    -9999                    -9999   
1                    -9999                    -9999                    -9999   
2                    -9999                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                    -9999                        0                    -9999   
1                    -9999                        0                    -9999   
2                    -9999                        0                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE050TI  
0                    -9999  
1                    -9999  
2                    -9999  
3                        0  
4                        0

In [5]:
climate_precip.shape

(151110, 29)

## merge()

In this section, you will learn about `merge()` functionality in Pandas.

### Inner Join

Here, you will use a plain `merge()` call to do an inner join and learn how this can result in a smaller, more focused dataset. First, you will create a new DataFrame object that contains the precipitation data from one station.

In [6]:
precip_one_station = climate_precip.query("STATION == 'GHCND:USC00045721'")
precip_one_station.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141                       86   

      DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
1460                       31                       11   
1461                       31                       11   
1462                       31                       11   
1463                       32                       11   
1464                       32                       11   

      DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
1460                        4                        0   
1461                        4                        0   
1462                        4                        0   
1463                        4                        0   
1464                        4                        0   

      DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
1460                        3                        3   
1461                        3                        3   
1462                        3                        3   
1463                        3                        2   
1464                        3                        2   

      DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
1460                        9                        6   
1461                       10                        6   
1462                       10                        6   
1463                       10                        6   
1464                       10                        6   

      DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
1460                        0                    -9999   
1461                        0                    -9999   
1462                        0                    -9999   
1463                        0                    -9999   
1464                        0                    -9999   

      DLY-SNOW-PCTALL-GE050TI  
1460                    -9999  
1461              

In [7]:
precip_one_station.shape

(365, 29)

In [8]:
inner_merged = pd.merge(precip_one_station, climate_temp)
inner_merged.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI ELEVATION LATITUDE  LONGITUDE  DLY-CLDD-BASE45  \
0                    -9999    1325.9  34.9436  -115.5469                3   
1                    -9999    1325.9  34.9436  -115.5469                3   
2                    -9999    1325.9  34.9436  -115.5469                3   
3            

How many rows do you think this merged DataFrame has?

In [9]:
inner_merged.shape

(365, 47)

You get 365 rows because any non-matching rows are discarded in an inner join, which is the default merge method for a `merge()` call, and `precip_one_station` had only 365 rows.

What if you want to merge both full datasets, but specify which columns to join on? In this case, you will use the `on` parameter:

In [10]:
inner_merged_total = pd.merge(
    climate_temp, climate_precip, on=["STATION", "DATE"]
)
inner_merged_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

In [11]:
inner_merged_total.shape

(123005, 48)

You can specify a single _key column_ with a string, or multiple key columns with a list, as in the above example. This results in a DataFrame with 123005 rows and 48 columns. 

Why 48 columns instead of 47? Because you specified the keys columns to join on, Pandas doesn't try to merge all mergeable columns. This can result in "duplicate" column names, which may or may not have different values. "Duplicate" is in quotes because the columns will actually have new names, by default they are appended with `_x` and `_y`. You can also use the `suffixes` parameter to control what is appended to the column names.

### Outer Join
With the outer join, you will retain rows that don't have matches as well. For this example, you will use the smaller precipitation DataFrame `precip_one_station` with the full `climate_temp` DataFrame and join with `STATION` and `DATE` columns as the key columns. Take a second and think about how many rows you expect the new DataFrame to have.

In [12]:
outer_merged = pd.merge(
    precip_one_station, climate_temp, how="outer", on=["STATION", "DATE"]
)
outer_merged.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0           -666.0            -66.6             0.16           -666.0   
1           -666.0            -66.6             0.16           -666.0   
2           -666.0            -66.6             0.16           -666.0   
3           -666.0            -66.6             0.16           -666.0   
4           -666.0            -66.6             0.17           -666.0   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44           -666.0            -66.6   
1            -66.6             0.44           -666.0            -66.6   
2            -66.6             0.45           -666.0            -66.6   
3            -66.6             0.45           -666.0            -66.6   
4            -66.6             0.46           -666.0            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                    137.0                     81.0                     31.0   
1                    138.0                     83.0                     31.0   
2                    139.0                     84.0                     31.0   
3                    140.0                     85.0                     32.0   
4                    141.0                     86.0                     32.0   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                     11.0                      4.0                      0.0   
1                     11.0                      4.0                      0.0   
2                     11.0                      4.0                      0.0   
3                     11.0                      4.0                      0.0   
4                     11.0                      4.0                      0.0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                      3.0                      3.0                      9.0   
1                      3.0                      3.0                     10.0   
2                      3.0                      3.0                     10.0   
3                      3.0                      2.0                     10.0   
4                      3.0                      2.0                     10.0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                      6.0                      0.0                  -9999.0   
1                      6.0                      0.0                  -9999.0   
2                      6.0                      0.0                  -9999.0   
3                      6.0                      0.0                  -9999.0   
4                      6.0                      0.0                  -9999.0   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION LATITUDE  \
0                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
1                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
2                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
3                  -9999.0  M

In [13]:
outer_merged.shape

(127020, 48)

If you remember from when you checked the `.shape` attribute of `climate_temp`, you'll see that the number of rows in `outer_merged` matches that. With an outer join, you can expect to have the same number of rows as the larger DataFrame, since none are lost like they are in an inner join. 

### Left Join
Also known as a left outer join. In this join, you will retain rows that don't have matches only on the left (or first) DataFrame to be merged.

In [14]:
left_merged = pd.merge(
    climate_temp, precip_one_station, how="left", on=["STATION", "DATE"]
)
left_merged.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

  STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0            NaN              NaN              NaN              NaN   
1            NaN              NaN              NaN              NaN   
2            NaN              NaN              NaN              NaN   
3            NaN              NaN              NaN              NaN   
4            NaN              NaN              NaN              NaN   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0              NaN              NaN                      NaN   
1              NaN              NaN                      NaN   
2              NaN              NaN                      NaN   
3              NaN              NaN                      NaN   
4              NaN              NaN                      NaN   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
0                      NaN                      

In [15]:
left_merged.shape

(127020, 48)

Here, you see that the number of rows in the resulting DataFrame matches that of the rows in the `climate_temp` DataFrame. What if we switched the positions of the two DataFrames that we are merging?

In [16]:
left_merged_reversed = pd.merge(
    precip_one_station, climate_temp, how="left", on=["STATION", "DATE"]
)
left_merged_reversed.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION LATITUDE  \
0                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
1                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
2                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
3                    -9999  M

In [17]:
left_merged_reversed.shape

(365, 48)

### Right Join
This works the same as the left join, however non-matching rows are only retained in the _right_ DataFrame. In the next example, you will recreate the `left_merged` DataFrame but with a right join.

In [18]:
right_merged = pd.merge(
    precip_one_station, climate_temp, how="right", on=["STATION", "DATE"]
)
right_merged.head()

STATION STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099            NaN  20100101              NaN   
1  GHCND:USC00049099            NaN  20100102              NaN   
2  GHCND:USC00049099            NaN  20100103              NaN   
3  GHCND:USC00049099            NaN  20100104              NaN   
4  GHCND:USC00049099            NaN  20100105              NaN   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION  LATITUDE  \
0                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
1                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
2                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
3                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
4    

In [19]:
right_merged.shape

(127020, 48)

Here, you simply flipped the positions of the input DataFrames and specified a right join. When you inspect `right_merged`, you might notice that it's not exactly the same as `left_merged`. The only difference between the two is the order of the columns: the first input's columns will always be the first in the newly formed DataFrame.

## .join()
`.join()` uses `merge()` under the hood, but provides a much more simplified interface to `merge()` and by default joins on indexes. Here is an introductory example using the `lsuffix` and `rsuffix` parameters to handle overlapping column names.

In [20]:
precip_one_station.join(climate_temp, lsuffix="_left", rsuffix="_right")

STATION_left       STATION_NAME_left  DATE_left  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100105             0.05   
...                 ...                     ...        ...              ...   
1820  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101227             0.04   
1821  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101228             0.04   
1822  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101229             0.04   
1823  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101230             0.04   
1824  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101231             0.04   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   
...               ...              ...              ...              ...   
1820             -666            -66.6             0.15             -666   
1821             -666            -66.6             0.15             -666   
1822             -666            -66.6             0.15             -666   
1823             -666            -66.6             0.15             -666   
1824             -666            -66.6             0.15             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   
...               ...              ...              ...              ...   
1820            -66.6             0.44             -666            -66.6   
1821            -66.6             0.43             -666            -66.6   
1822            -66.6             0.43             -666            -66.6   
1823            -66.6             0.43             -666            -66.6   
1824            -66.6             0.44             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   
...               ...              ...              ...              ...   
1820             1.14              0.3            11.29              1.5   
1821             1.19              0.3            11.34              1.5   
1822             1.24              0.3            11.39              1.5   
1823             1.30              0.3            11.45              1.5   
1824             1.35              0.3            11.50              1.5   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141    

If you inspect the data, you'll see that overlapping columns are kept, just renamed to be unique. If we flip this around and instead call `.join()` on the larger DataFrame, you'll notice that the DataFrame is larger, but data that doesn't exist in the smaller DataFrame (`precip_one_station`) is filled in with `NaN` (_Not a Number_) values.

In [21]:
climate_temp.join(precip_one_station, lsuffix="_left", rsuffix="_right")

STATION_left       STATION_NAME_left ELEVATION  LATITUDE  \
0       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
1       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
2       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
3       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
4       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
...                   ...                     ...       ...       ...   
127015  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127016  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127017  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127018  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127019  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   

         LONGITUDE  DATE_left  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
0       -116.03694   20100101                6                2   
1       -116.03694   20100102                6                2   
2       -116.03694   20100103                6                2   
3       -116.03694   20100104                6                2   
4       -116.03694   20100105                6                2   
...            ...        ...              ...              ...   
127015   -118.0711   20101227                4                2   
127016   -118.0711   20101228                4                2   
127017   -118.0711   20101229                4                2   
127018   -118.0711   20101230                4                2   
127019   -118.0711   20101231                4                2   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
0                 -7777            -7777            -7777                0   
1                     1            -7777            -7777                0   
2                     1            -7777            -7777                0   
3                     1            -7777            -7777                0   
4                     1            -7777            -7777                0   
...                 ...              ...              ...              ...   
127015                1            -7777            -7777            -7777   
127016                1            -7777            -7777            -7777   
127017                1            -7777            -7777            -7777   
127018                1            -7777            -7777            -7777   
127019                1            -7777            -7777            -7777   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
0                     0                0            -7777                1   
1                     0                0            -7777                1   
2                     0                0            -7777                1   
3                     0                0            -7777                1   
4                     0                0            -7777            -7777   
...                 ...              ...              ...              ...   
127015                0                0                2                4   
127016                0                0                2                3   
127017                0                0                2                3   
127018                0                0                2                3   
127019                0                0                1                3   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
0                     2                6                7               10   
1                     2                6                7               10   
2                     2                5                7               10   
3                     2                5                7               10   
4                     2                5                7          

If you must use `.join()` and want to merge the columns, you must set them to be indexes first. First take a look at this previously used `merge()` operation:

In [22]:
inner_merged_total = pd.merge(
    climate_temp, climate_precip, on=["STATION", "DATE"]
)
inner_merged_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

In [23]:
inner_joined_total = climate_temp.join(
    climate_precip.set_index(["STATION", "DATE"]),
    on=["STATION", "DATE"],
    how="inner",
    lsuffix="_x",
    rsuffix="_y",
)
inner_joined_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

Because `.join()` works on indexes, if we want to recreate `merge()` before, then we must set indexes on the join columns we specify. In this example, you used the `.set_index()` method to set your indexes to the key columns within the join. 

Below you will see an almost-bare `.join()` call. Because there are overlapping columns, you will have to specify a suffix with `lsuffix`, `rsuffix`, or both, but this example will demonstrate the more typical behavior of `.join()`.

In [24]:
climate_temp.join(climate_precip, lsuffix="_left")

STATION_left       STATION_NAME_left ELEVATION  LATITUDE  \
0       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
1       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
2       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
3       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
4       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
...                   ...                     ...       ...       ...   
127015  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127016  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127017  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127018  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127019  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   

         LONGITUDE  DATE_left  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
0       -116.03694   20100101                6                2   
1       -116.03694   20100102                6                2   
2       -116.03694   20100103                6                2   
3       -116.03694   20100104                6                2   
4       -116.03694   20100105                6                2   
...            ...        ...              ...              ...   
127015   -118.0711   20101227                4                2   
127016   -118.0711   20101228                4                2   
127017   -118.0711   20101229                4                2   
127018   -118.0711   20101230                4                2   
127019   -118.0711   20101231                4                2   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
0                 -7777            -7777            -7777                0   
1                     1            -7777            -7777                0   
2                     1            -7777            -7777                0   
3                     1            -7777            -7777                0   
4                     1            -7777            -7777                0   
...                 ...              ...              ...              ...   
127015                1            -7777            -7777            -7777   
127016                1            -7777            -7777            -7777   
127017                1            -7777            -7777            -7777   
127018                1            -7777            -7777            -7777   
127019                1            -7777            -7777            -7777   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
0                     0                0            -7777                1   
1                     0                0            -7777                1   
2                     0                0            -7777                1   
3                     0                0            -7777                1   
4                     0                0            -7777            -7777   
...                 ...              ...              ...              ...   
127015                0                0                2                4   
127016                0                0                2                3   
127017                0                0                2                3   
127018                0                0                2                3   
127019                0                0                1                3   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
0                     2                6                7               10   
1                     2                6                7               10   
2                     2                5                7               10   
3                     2                5                7               10   
4                     2                5                7          

## concat()

First, you will see a basic concatenation along axis 0.

In [25]:
double_precip = pd.concat([precip_one_station, precip_one_station])
double_precip.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141                       86   

      DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
1460                       31                       11   
1461                       31                       11   
1462                       31                       11   
1463                       32                       11   
1464                       32                       11   

      DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
1460                        4                        0   
1461                        4                        0   
1462                        4                        0   
1463                        4                        0   
1464                        4                        0   

      DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
1460                        3                        3   
1461                        3                        3   
1462                        3                        3   
1463                        3                        2   
1464                        3                        2   

      DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
1460                        9                        6   
1461                       10                        6   
1462                       10                        6   
1463                       10                        6   
1464                       10                        6   

      DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
1460                        0                    -9999   
1461                        0                    -9999   
1462                        0                    -9999   
1463                        0                    -9999   
1464                        0                    -9999   

      DLY-SNOW-PCTALL-GE050TI  
1460                    -9999  
1461              

To reset the index, use the `ignore_index` optional parameter.

In [26]:
reindexed = pd.concat(
    [precip_one_station, precip_one_station], ignore_index=True
)
reindexed.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI  
0                    -9999  
1                    -9999  
2                    -9999  
3                    -9999  
4                    -9999

When axis labels for the axis you are **not** concatenating along don't match (for example, column labels when concatenating along rows), then all columns are preserved and missing data is filled in with `NaN`. 

In [27]:
outer_joined = pd.concat([climate_precip, climate_temp])
outer_joined.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0           -666.0            -66.6            -6.66           -666.0   
1           -666.0            -66.6            -6.66           -666.0   
2           -666.0            -66.6            -6.66           -666.0   
3          -9999.0          -9999.0            -6.66          -9999.0   
4          -9999.0          -9999.0            -6.66          -9999.0   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6            -6.66           -666.0            -66.6   
1            -66.6            -6.66           -666.0            -66.6   
2            -66.6            -6.66           -666.0            -66.6   
3          -9999.0            -6.66          -9999.0          -9999.0   
4          -9999.0            -6.66          -9999.0          -9999.0   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.02              0.0             0.02              0.0   
1             0.04              0.0             0.04              0.0   
2             0.05              0.0             0.05              0.0   
3             0.07              0.0             0.07              0.0   
4             0.09              0.0             0.09              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                     98.0                     43.0                     12.0   
1                     99.0                     44.0                     12.0   
2                    100.0                     44.0                     12.0   
3                    101.0                     45.0                     12.0   
4                    102.0                     45.0                     12.0   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                      3.0                  -9999.0                      0.0   
1                      3.0                  -9999.0                      0.0   
2                      3.0                  -9999.0                      0.0   
3                      3.0                      0.0                      0.0   
4                      3.0                      0.0                      0.0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                  -9999.0                  -9999.0                  -9999.0   
1                  -9999.0                  -9999.0                  -9999.0   
2                  -9999.0                  -9999.0                  -9999.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                  -9999.0                      0.0                  -9999.0   
1                  -9999.0                      0.0                  -9999.0   
2                  -9999.0                      0.0                  -9999.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   DLY-SNOW-PCTALL-GE050TI ELEVATION LATITUDE LONGITUDE  DLY-CLDD-BASE45  \
0                  -9999.0       NaN      NaN       NaN              NaN   
1                  -9999.0       NaN      NaN       NaN              NaN   
2                  -9999.0       NaN      NaN       NaN              NaN   
3                

In [28]:
inner_joined = pd.concat([climate_temp, climate_precip], join="inner")
inner_joined.head()

STATION            STATION_NAME      DATE
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105

In [29]:
inner_joined.shape

(278130, 3)

To illustrate how this would work with rows, concatenate along columns instead:

In [30]:
inner_joined_cols = pd.concat(
    [climate_temp, climate_precip], axis="columns", join="inner"
)
inner_joined.head()

STATION            STATION_NAME      DATE
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105

In [31]:
inner_joined_cols.shape

(127020, 50)

You can also use the `keys` parameter to set hierarchical axis labels which can be used, for example, to preserve original labels while maintaining labels that tell you which dataset each row or column came from.

In [32]:
hierarchical_keys = pd.concat(
    [climate_temp, climate_precip], keys=["temp", "precip"]
)
hierarchical_keys.head()

STATION            STATION_NAME ELEVATION  LATITUDE  \
temp 0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   

         LONGITUDE      DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
temp 0  -116.03694  20100101              6.0              2.0   
     1  -116.03694  20100102              6.0              2.0   
     2  -116.03694  20100103              6.0              2.0   
     3  -116.03694  20100104              6.0              2.0   
     4  -116.03694  20100105              6.0              2.0   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
temp 0          -7777.0          -7777.0          -7777.0              0.0   
     1              1.0          -7777.0          -7777.0              0.0   
     2              1.0          -7777.0          -7777.0              0.0   
     3              1.0          -7777.0          -7777.0              0.0   
     4              1.0          -7777.0          -7777.0              0.0   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
temp 0              0.0              0.0          -7777.0              1.0   
     1              0.0              0.0          -7777.0              1.0   
     2              0.0              0.0          -7777.0              1.0   
     3              0.0              0.0          -7777.0              1.0   
     4              0.0              0.0          -7777.0          -7777.0   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
temp 0              2.0              6.0              7.0             10.0   
     1              2.0              6.0              7.0             10.0   
     2              2.0              5.0              7.0             10.0   
     3              2.0              5.0              7.0             10.0   
     4              2.0              5.0              7.0             10.0   

        DLY-HTDD-NORMAL  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
temp 0             15.0              NaN              NaN              NaN   
     1             15.0              NaN              NaN              NaN   
     2             15.0              NaN              NaN              NaN   
     3             15.0              NaN              NaN              NaN   
     4             15.0              NaN              NaN              NaN   

        DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
temp 0              NaN              NaN              NaN              NaN   
     1              NaN              NaN              NaN              NaN   
     2              NaN              NaN              NaN              NaN   
     3              NaN              NaN              NaN              NaN   
     4              NaN              NaN              NaN              NaN   

        DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
temp 0              NaN              NaN              NaN              NaN   
     1              NaN              NaN              NaN              NaN   
     2              NaN              NaN              NaN              NaN   
     3              NaN              NaN              NaN              NaN   
     4              NaN              NaN              NaN              NaN   

        YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
temp 0              NaN              NaN                      NaN   
     1              NaN              NaN                      NaN   
     2              NaN              NaN                      NaN   
     3              NaN              NaN                      NaN   
     4              NaN

In [33]:
hierarchical_keys.tail()

STATION            STATION_NAME ELEVATION LATITUDE  \
precip 151105  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151106  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151107  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151108  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151109  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   

              LONGITUDE      DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
precip 151105       NaN  20101227              NaN              NaN   
       151106       NaN  20101228              NaN              NaN   
       151107       NaN  20101229              NaN              NaN   
       151108       NaN  20101230              NaN              NaN   
       151109       NaN  20101231              NaN              NaN   

               DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-CLDD-NORMAL  DLY-CLDD-BASE70  DLY-CLDD-BASE72  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-NORMAL  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  \
precip 151105              NaN             0.12          -9999.0   
       151106              NaN             0.12          -9999.0   
       151107              NaN             0.12          -9999.0   
       151108              NaN             0.12          -9999.0   
       151109              NaN             0.12          -9999.0   

               DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
precip 151105          -9999.0             0.41          -9999.0   
       151106          -9999.0             0.42          -9999.0   
       151107          -9999.0             0.43          -9999.0   
       151108          -9999.0             0.43          -9999.0   
       151109          -9999.0             0.44          -9999.0   

               DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  \
precip 151105          -9999.0             1.13          -9999.0   
       151106          -9999.0             1.14          -9999.0   
       151107          -9999.0             1.15          -9999.0   
       151108          -9999.0             1.15          -9999.0   
       151109          -9999.0             1.16          -9999.0   

               DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
precip 151105          -9999.0             3.91          -9999.0   
       151106          -9999.0             4.10          -9999.0   
       151107          -9999.0             4.30         